File: python/leetcode/205.py
class Solution(object):
    def isIsomorphic(self, s, t):
        if len(s) != len(t):
            return False
        word1, word2 = {}, {}
        for i in range(len(s)):
            ch1, ch2 = s[i], t[i]
            if word1.get(ch1, 0) != word2.get(ch2, 0):
                return False
            word1[ch1] = i + 1
            word2[ch2] = i + 1
        return True


class Solution(object):
    def isIsomorphic(s, t):
        if len(s) != len(t):
            return False
        word = {}
        for i in range(len(s)):
            ch1, ch2 = s[i], t[i]
            if ch1 in word:
                if word[ch1] != ch2:
                    return False
            else:
                if ch2 in word.values():
                    return False
                word[ch1] = ch2
        return True

File: cpp/leetcode/205.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }
        int word1[256] = {0}, word2[256] = {0}, n = s.size();
        for (int i = 0; i < n; ++i) {
            if (word1[s[i]] != word2[t[i]]) {
                return false;
            }
            word1[s[i]] = i + 1; // avoid 0 idx
            word2[t[i]] = i + 1;
        }
        return true;
    }
};

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }
        unordered_map<char, int> smap;
        unordered_map<char, int> tmap;
        for (int i = 0; i < s.size(); i++) {
            int sidx = smap.find(s[i]) != smap.end() ? smap[s[i]] : -1;
            int tidx = tmap.find(t[i]) != tmap.end() ? tmap[t[i]] : -1;
            if (sidx != tidx) {
                return false;
            }
            smap[s[i]] = i;
            tmap[t[i]] = i;
        }
        return true;
    }
};